src/qibolab/instruments/qm.py

```diff
@@ -782,15 +842,21 @@ def create_qmsequence(self, qubits, sequence):
         # like we do for readout multiplex
         qmsequence = Sequence()
         for pulse in sorted(sequence.pulses, key=lambda pulse: (pulse.start, pulse.duration)):
-            qmpulse = qmsequence.add(pulse)
             if pulse.type is PulseType.FLUX:
                 # register flux element (if it does not already exist)
                 self.config.register_flux_element(qubits[pulse.qubit], pulse.frequency)
-                qmpulse.bake(self.config)
+                qmpulse = BakedPulse(pulse)
+                qmpulse.bake(self.config, durations=[pulse.duration])
+                qmsequence.add(qmpulse)
             else:
+                qmpulse = QMPulse(pulse)
+                qmsequence.add(qmpulse)
                 if pulse.duration % 4 != 0 or pulse.duration < 16:
                     raise_error(NotImplementedError, "1ns resolution is available for flux pulses only.")
```

Comment:
The `qmsequence.add(qmpulse)` part is common to the `if` and `else` branches, so maybe you can put just right after.

The only reason for which is not that straightforward is that it's not the last instruction in the `else` branch, but the only one after is raising an error, so there should be no problem in swapping the two (in principle, you could also raise the error immediately at the beginning of the `else` branch, and save some computation in case you would raise).